src/Sparrow.Server/Platform/MemoryLockUsage.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Sparrow.Binary;
+using Sparrow.Global;
+using Sparrow.Platform;
+
+namespace Sparrow.Server.Platform;
+
+public static unsafe class MemoryLockUsage
+{
+        public static long LockedMemorySize;
+        
+        public static event EventHandler<long> MemoryLockedCalled;
+
+        [UnmanagedCallersOnly(CallConvs = [typeof(CallConvCdecl)])]
+        public static void UpdateLockedMemory(Int64 v, char* filenamePtr)
+        {
+            MemoryLockedCalled?.Invoke(null, v);
+            Interlocked.Add(ref LockedMemorySize, v);
+        }
+        
+        [UnmanagedCallersOnly(CallConvs = [typeof(CallConvCdecl)])]
+        public static int RecoverLockedMemoryFailure(Int64 sizeToLock, char* filenamePtr)
+        {
+            if (PlatformDetails.RunningOnPosix)
+                return 0; // nothing to do here
+            using var currentProcess = Process.GetCurrentProcess();
+            var nextWorkingSetSize = GetNearestFileSize(currentProcess.MinWorkingSet.ToInt64() + sizeToLock);
+            if (nextWorkingSetSize > int.MaxValue && PlatformDetails.Is32Bits)
+            {
+                nextWorkingSetSize = int.MaxValue;
+            }
+            
+#pragma warning disable CA1416 // Validate platform compatibility
+            // Minimum working set size must be less than or equal to the maximum working set size.
+            // Let's increase the max as well.
+            if (nextWorkingSetSize > currentProcess.MaxWorkingSet)
+            {
+                try
+                {
+                    currentProcess.MaxWorkingSet = new IntPtr(nextWorkingSetSize);
+                }
+                catch
+                {
+                    // cannot throw, called from native code
+                    return 0;
+                }
+            }
+            
+            try
+            {
+                currentProcess.MinWorkingSet = new IntPtr(nextWorkingSetSize);
+            }
+            catch
+            {
+                // cannot throw, called from native code
+                return 0;
+            }
+#pragma warning restore CA1416 // Validate platform compatibility
+
+            return 1;
+        }
+        
+        private static readonly long IncreaseByPowerOf2Threshold = new Size(512, SizeUnit.Megabytes).GetValue(SizeUnit.Bytes);
+        private static long GetNearestFileSize(long neededSize)
+        {
+            if (neededSize < IncreaseByPowerOf2Threshold)
+                return Bits.PowerOf2(neededSize);
```

Comment:
should this be Bits.NextAllocationSize ?

Reply:
Probably

Reply:
It's okay here. We have `if (neededSize < IncreaseByPowerOf2Threshold)` so we grow by power of 2 only below 512MB